doc/graph.md

```diff
@@ -1,3 +1,399 @@
+// The output is: x=1.98708
+```
+### Debugging
+For debugging, we can call `debug()` to print node parameters. The `debug()` function has to be called prior to graph execution. Once a node is marked for debugging, its value (resulting tensor) and gradient will be printed out during the forward and backward pass. It is also recommended to turn on Marian logger by calling `createLoggers()` for more information.
+
+**Debugging for modified Example 1**
+
+Suppose we want to check the results of node `x` during the computation. We can call `debug()` to mark node `x` for debugging. 
+```cpp
+// mark node x for debugging with logging message "Parameter x"
+debug(x, "Parameter x");
+```
+The output is shown as follows with `createLoggers()`:
+```cpp
+[2021-02-16 15:10:51] [memory] Reserving 256 B, device gpu0
+[2021-02-16 15:10:51] Debug: Parameter x op=param
+[2021-02-16 15:10:51] shape=1x1 size=1 type=float32 device=gpu0 ptr=140505547538432 bytes=256
+min: 2.00000000 max: 2.00000000 l2-norm: 2.00000000
+[[   2.00000000 ]]
+
+[2021-02-16 15:10:51] [memory] Reserving 256 B, device gpu0
+[2021-02-16 15:10:51] Debug Grad: Parameter x op=param
+[2021-02-16 15:10:51] shape=1x1 size=1 type=float32 device=gpu0 ptr=140505547538944 bytes=256
+min: 2.58385324 max: 2.58385324 l2-norm: 2.58385324
+[[   2.58385324 ]]
+```
+
+### More advanced
+
+For more details about graph execution, a graph keeps track of all the `Node` objects in its `nodesForward_` and `nodesBackward_` lists. 
+`nodesForward_` contains all nodes used for the forward pass and `nodesBackward_` contains all trainable nodes used for the backward pass. 
+All the tensor objects for a graph is stored in its `tensors_` attribute. 
+`tensors_` is a shared pointer holding memory and nodes for a graph. 
+Since each `Node` can result in new tensors, this attribute is used to allocate memory for new tensors during the forward and backward pass.
+This `tensors_` attribute gets cleared before a new graph is built. 
+Another important attribute in `ExpressionGraph` is `paramsByElementType_`. 
+This attribute holds memory and nodes that corresponds to graph parameters.
+You can call `params()` function in a graph to get all the parameter objects:
+
+```cpp
+// return the Parameters object related to the graph
+// The Parameters object holds the whole set of the parameter nodes.
+graph->params();
+```
+
+Besides, we provide APIs to support the mechanism of Gradient Checkpointing. 
+This method works by trading compute for memory, which reruns a forward-pass segment for each checkpoint segment during backward pass. 
+Currently, Marian only support set checkpoint nodes manually (by calling `Node::markCheckpoint()` or `checkpoint()`). 
+To enable the mode of gradient-checkpointing for a graph, we use `setCheckpointing()`:
+
+```cpp
+// enable gradient-checkpointing for a graph
+graph->setCheckpointing(true);
+```
+
+We can also save and load the model (containing all the parameters) in Marian. 
```

Comment:
The model is graph + parameters?

Reply:
The model here only contains the parameters.